scripts/external_code.py

```diff
@@ -1,453 +1 @@
-
-    cn_script_args_diff = 0
-    for script in script_runner.alwayson_scripts:
-        if script is cn_script:
-            cn_script_args_diff = len(cn_units) - (cn_script.args_to - cn_script.args_from)
-            script_args[script.args_from:script.args_to] = cn_units
-            script.args_to = script.args_from + len(cn_units)
-        else:
-            script.args_from += cn_script_args_diff
-            script.args_to += cn_script_args_diff
-
-
-def get_models(update: bool=False) -> List[str]:
-    """
-    Fetch the list of available models.
-    Each value is a valid candidate of `ControlNetUnit.model`.
-
-    Keyword arguments:
-    update -- Whether to refresh the list from disk. (default False)
-    """
-
-    if update:
-        global_state.update_cn_models()
-
-    return list(global_state.cn_models_names.values())
-
-
-def get_modules(alias_names: bool = False) -> List[str]:
-    """
-    Fetch the list of available preprocessors.
-    Each value is a valid candidate of `ControlNetUnit.module`.
-
-    Keyword arguments:
-    alias_names -- Whether to get the ui alias names instead of internal keys
-    """
-
-    modules = list(global_state.cn_preprocessor_modules.keys())
-
-    if alias_names:
-        modules = [global_state.preprocessor_aliases.get(module, module) for module in modules]
-
-    return modules
-
-
-def get_modules_detail(alias_names: bool = False) -> Dict[str, Any]:
-    """
-    get the detail of all preprocessors including
-    sliders: the slider config in Auto1111 webUI
-
-    Keyword arguments:
-    alias_names -- Whether to get the module detail with alias names instead of internal keys
-    """
-
-    _module_detail = {}
-    _module_list = get_modules(False)
-    _module_list_alias = get_modules(True)
-    
-    _output_list = _module_list if not alias_names else _module_list_alias
-    for index, module in enumerate(_output_list):
-        if _module_list[index] in preprocessor_sliders_config:
-            _module_detail[module] = {
-                "model_free": module in model_free_preprocessors,
-                "sliders": preprocessor_sliders_config[_module_list[index]]
-            }
-        else:
-            _module_detail[module] = {
-                "model_free": False,
-                "sliders": []
-            }
-            
-    return _module_detail
-
-
-def find_cn_script(script_runner: scripts.ScriptRunner) -> Optional[scripts.Script]:
-    """
-    Find the ControlNet script in `script_runner`. Returns `None` if `script_runner` does not contain a ControlNet script.
-    """
-
-    if script_runner is None:
-        return None
-
-    for script in script_runner.alwayson_scripts:
-        if is_cn_script(script):
-            return script
-
-
-def is_cn_script(script: scripts.Script) -> bool:
-    """
-    Determine whether `script` is a ControlNet script.
-    """
-
-    return script.title().lower() == 'controlnet'
+from internal_controlnet.external_code import *
```

Comment:
Do we keep `scripts/external_code.py` here for compatibility. Do other extensions import this module directly?

Reply:
Yes, that is the idea. SD-CN-animation, deforum and other extensions import this module at this path so we cannot remove it completely. I guess a comment in the code would make this clear